pkgs/README: recommend & redirect to pkgs/by-name in versioning section (#334413)
NixOS/nixpkgs: pkgs/README.md

Because every version of a package in Nixpkgs creates a potential maintenance burden, old versions of a package should not be kept unless there is a good reason to do so. For instance, Nixpkgs contains several versions of GCC because other packages don’t build with the latest version of GCC. Other examples are having both the latest stable and latest pre-release version of a package, or to keep several major releases of an application that differ significantly in functionality.

> If there is only one version of a package, its Nix expression should be named (e.g) `pkgs/by-name/xy/xyz/package.nix`. If there are multiple versions, this should be reflected in the attribute name. If you wish to share code between the Nix expressions of each version, you won't have a choice, but to not rely upon `pkgs/by-name`'s automatic attribute creation, and create the attributes yourself in `all-paclages.nix`. See also [`pkgs/by-name/README.md`'s section about the topic](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md#recommendation-for-new-packages-with-multiple-versions).

Reviewer: ```suggestion
If there is only one version of a package, its Nix expression should be named (e.g) `pkgs/by-name/xy/xyz/package.nix`. If there are multiple versions, this should be reflected in the attribute name. If you wish to share code between the Nix expressions of each version, you cannot rely upon `pkgs/by-name`'s automatic attribute creation, and must create the attributes yourself in `all-paclages.nix`. See also [`pkgs/by-name/README.md`'s section on this topic](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md#recommendation-for-new-packages-with-multiple-versions).
```

> If there is only one version of a package, its Nix expression should be named (e.g) `pkgs/by-name/xy/xyz/package.nix`. If there are multiple versions, this should be reflected in the attribute name (which is identical to the name of the directory holding the versioned `package.nix` file). E.g. `pkgs/by-name/xy/xyz_1_41_8/package.nix` and `pkgs/by-name/xy/xyz_1_41_9/package.nix`. The version in the directory should leave out unnecessary detail. For instance, if we keep the latest Firefox `2.0.x` and `3.5.x` versions in Nixpkgs, they should be named `pkgs/by-name/fi/firefox_2_0/package.nix` and `pkgs/by-name/fi/firefox_3_5/package.nix`, respectively (which, at a given point, might contain versions `2.0.0.20` and `3.5.4`).

Reviewer 2: I think the Firefox example could be shortened to firefox_2 and firefox_3. Otherwise LGTM
